File: tests/ut/cpp/ops/test_ops_squeeze.cc
/**
 * Copyright 2023 Huawei Technologies Co., Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <vector>
#include <memory>
#include "common/common_test.h"
#include "ir/dtype/type.h"
#include "abstract/dshape.h"
#include "utils/tensor_construct_utils.h"
#include "ir/primitive.h"
#include "abstract/abstract_value.h"
#include "include/backend/optimizer/helper.h"
#include "ops/test_ops.h"
#include "infer/ops_func_impl/squeeze.h"
#include "ops/test_value_utils.h"

namespace mindspore {
namespace ops {
#define I64(x) (static_cast<int64_t>((x)))

struct SqueezeShapeParams {
  ShapeVector input_x_shape;
  TypePtr input_x_type;
  ValuePtr axis;
  ShapeVector output_shape;
  TypePtr output_type;
};

class TestSqueeze : public TestOps, public testing::WithParamInterface<SqueezeShapeParams> {};

TEST_P(TestSqueeze, dyn_shape) {
  const auto &param = GetParam();
  auto input_x = std::make_shared<abstract::AbstractTensor>(param.input_x_type, param.input_x_shape);
  auto expect_shape = std::make_shared<abstract::Shape>(param.output_shape);
  auto expect_dtype = std::make_shared<TensorType>(param.output_type);

  SqueezeFuncImpl squeeze_func_impl;
  auto prim = std::make_shared<Primitive>("Squeeze");

  auto out_dtype = squeeze_func_impl.InferType(prim, {input_x, param.axis->ToAbstract()});
  ASSERT_TRUE(*out_dtype == *expect_dtype);
  auto out_shape = squeeze_func_impl.InferShape(prim, {input_x, param.axis->ToAbstract()});
  ASSERT_TRUE(*out_shape == *expect_shape);
}

INSTANTIATE_TEST_CASE_P(
  TestSqueeze, TestSqueeze,
  testing::Values(SqueezeShapeParams{{3, 4, 1, 5}, kFloat32, CreateTuple({I64(2)}), {3, 4, 5}, kFloat32},
                  SqueezeShapeParams{{3, 4, 1, 5, 1}, kFloat32, CreateTuple({I64(2), I64(4)}), {3, 4, 5}, kFloat32},
                  SqueezeShapeParams{{1, 3, 4, 5}, kInt64, CreateTuple({I64(0)}), {3, 4, 5}, kInt64},
                  SqueezeShapeParams{{3, 4, 1, 5, 1}, kInt64, CreateTuple({I64(-3), I64(-1)}), {3, 4, 5}, kInt64},
                  SqueezeShapeParams{{3, 1, 4, 5}, kInt64, CreateTuple({I64(-3)}), {3, 4, 5}, kInt64},
                  SqueezeShapeParams{{2, 3, 4, 1, 5}, kInt32, CreateTuple({I64(-2)}), {2, 3, 4, 5}, kInt32},
                  SqueezeShapeParams{{-1, -1, -1, -1}, kUInt64, CreateTuple({I64(2)}), {-2}, kUInt64},
                  SqueezeShapeParams{
                    {-1, -1, -1, -1}, kFloat32, CreateTuple({I64(0), I64(1), I64(-1)}), {-2}, kFloat32},
                  SqueezeShapeParams{{-2}, kFloat32, CreateTuple({I64(2)}), {-2}, kFloat32}));
}  // namespace ops
}  // namespace mindspore

File: tests/st/ops/test_ops_squeeze.py
# Copyright 2024 Huawei Technologies Co., Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============================================================================
import pytest
import numpy as np
import mindspore as ms
from mindspore import ops, Tensor
from mindspore.mint import squeeze
from tests.st.utils import test_utils
from tests.mark_utils import arg_mark
from tests.st.ops.test_tools.test_op import TEST_OP


def generate_random_input(shape, dtype):
    return np.random.randn(*shape).astype(dtype)

@test_utils.run_with_cell
def squeeze_forward_func(x, dim):
    return squeeze(x, dim)


@test_utils.run_with_cell
def squeeze_backward_func(x, dim):
    return ms.grad(squeeze_forward_func, (0))(x, dim)


@test_utils.run_with_cell
def squeeze_vmap_func(x, dim):
    return ops.vmap(squeeze_forward_func, in_axes=(0, None), out_axes=0)(x, dim)


@arg_mark(plat_marks=['platform_ascend'], level_mark='level1', card_mark='onecard', essential_mark='essential')
@pytest.mark.parametrize('context_mode', [ms.GRAPH_MODE, ms.PYNATIVE_MODE])
@test_utils.run_test_with_On
def test_ops_squeeze_normal(context_mode):
    """
    Feature: pyboost function.
    Description: test function squeeze forward and backward.
    Expectation: expect correct result.
    """
    ms.set_context(jit_level='O0')
    ms.context.set_context(mode=context_mode)
    # forward
    # int
    x = Tensor(np.array([[[2, 2], [2, 2]]]), ms.float32)
    dim = 0
    out = squeeze_forward_func(x, dim)
    expect = [[2., 2.], [2., 2.]]
    np.testing.assert_allclose(out.asnumpy(), expect, rtol=1e-3)
    # tuple(ints)
    x1 = Tensor(np.array([[[[2, 2]], [[2, 2]]]]), ms.float32)
    dim1 = (0, -2)
    out1 = squeeze_forward_func(x1, dim1)
    expect1 = [[2., 2.], [2., 2.]]
    np.testing.assert_allclose(out1.asnumpy(), expect1, rtol=1e-3)

    # backward
    output_b = squeeze_backward_func(ms.Tensor(x), dim)
    expect_b = [[[1., 1.], [1., 1.]]]
    np.testing.assert_allclose(output_b.asnumpy(), expect_b, rtol=1e-3)


@arg_mark(plat_marks=['platform_ascend', 'platform_ascend910b'],
          level_mark='level1', card_mark='onecard', essential_mark='unessential')
@pytest.mark.parametrize('context_mode', [ms.GRAPH_MODE, ms.PYNATIVE_MODE])
@test_utils.run_test_with_On
def test_ops_squeeze_vmap(context_mode):
    """
    Feature: pyboost function.
    Description: test function squeeze vmap feature.
    Expectation: expect correct result.
    """
    ms.set_context(jit_level='O0')
    ms.context.set_context(mode=context_mode)
    x = Tensor(np.array([[[[[2., 2.], [2., 2.]]]]]), ms.float32)
    dim = 0
    output = squeeze_vmap_func(x, dim)
    expect = [[[[2., 2.], [2., 2.]]]]
    np.testing.assert_allclose(output.asnumpy(), expect, rtol=1e-3)


@arg_mark(plat_marks=['platform_ascend', 'platform_ascend910b'],
          level_mark='level1', card_mark='onecard', essential_mark='unessential')
def test_ops_squeeze_dynamic_shape():
    """
    Feature: pyboost function.
    Description: test function squeeze with dynamic shape and dynamic rank.
    Expectation: return the correct value.
    """
    x1 = generate_random_input((1, 2, 3, 4, 5), np.float32)
    dim1 = 0
    x2 = generate_random_input((4, 1, 5), np.float32)
    dim2 = 1
    TEST_OP(squeeze_forward_func, [[ms.Tensor(x1), dim1], [ms.Tensor(x2), dim2]],
            disable_mode=["GRAPH_MODE_GE"],
            disable_case=['ScalarTensor'])
